Extract the search component into `@automattic/search`

sarayourfriend opened the pull request:
#### Rationale

We need to extract this component so that it can be used inside of the language-picker component. I've also moved it outside of `@automattic/components` for the same reasons laid out here: https://github.com/Automattic/wp-calypso/pull/46020#issuecomment-707869250

#### Changes proposed in this Pull Request

* Extract the search component from Calypso into a separate package
* Removes analytics from the component. Consumers will need to implement their own using the appropriate hooks.

This is more or less a 1-to-1 reimplementation of the existing component. There are some differences, however:
* Updated ref usage
* Using WordPress components rather than Calypso components and icons
* Using TypeScript and fixing some potential coding errors
* Add a wrapping noop `form` to cause iOS to display a "Search" button rather than "return"

<img height="500" alt="iOS displaying a Search button instead of Return on the soft keyboard" src="https://user-images.githubusercontent.com/24264157/96284364-545db000-0f92-11eb-92a0-1351d1425d9d.jpeg"></img>


#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Run the new storybook command `yarn run search:storybook:start` and ensure the examples work as expected and cover the different use cases

Note: I didn't port over the existing tests because they aren't really that useful of tests.

#### Questions

1. The way this component is written it can go from controlled to uncontrolled (if you start to pass in value later in the component lifecycle. We could vastly simplify this component if we turned it into a fully uncontrolled component. For example, if we did this, we could remove the derived state logic and the "responsiveness" of the component to certain props changing (like do we really need to be able to handle changes in `delaySearch`? That seems like a waste and a supremely unlikely edge case usage). Shall we move forward with turning this component into an uncontrolled component (i.e., by not responding to changes in "static" props and not allowing `value` to be passed in?)

Fixes #

sarayourfriend committed:
Try extracting search component

sarayourfriend committed:
Clean types and remove complexities around debouncing

sarayourfriend committed:
Fix searching

sarayourfriend committed:
Clean up comments and deps

sarayourfriend committed:
Replace mistakingly removed styles

sarayourfriend committed:
Add i18n documentation

sarayourfriend committed:
Suppress type errors

a8ci18n commented on packages/search/src/search.tsx:
```diff
@@ -0,0 +1,417 @@
+					tabIndex={ enableOpenIcon ? 0 : undefined }
+					onKeyDown={ enableOpenIcon ? this.openListener : undefined }
+					aria-controls={ 'search-component-' + this.instanceId }
+					aria-label={ __( 'Open Search' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **31** times:
`translate( 'Open Search', { context: 'button label'} )` *ES Score: 14*

sarayourfriend committed:
Update yarn.lock

tyxla commented on packages/search/src/search.tsx:
```diff
@@ -0,0 +1,417 @@
+	};
+
+	UNSAFE_componentWillReceiveProps( nextProps: Props ) {
+		if ( this.props.isOpen !== nextProps.isOpen ) {
```
This feels a bit flawed. The component state can be changed by user intent but could be overridden by another prop change, feels a bit off. I'd expect `isOpen` to be used only for the initial state. Is there a real use case where this behavior is necessary?

saramarcondes replied:
> These cases would maybe be covered if the search component auto-closed when it loses focus -- means that another UI element is now active.

For this to work we'd need to always auto-focus. I'm not opposed to that given that a similar component in the new P2 follows this behavior but my understanding is that auto-focusing is undesirable for accessibility reasons. Maybe that doesn't apply for when you're "auto-focusing" on a new element after activating a button that makes the new element appear?

tyxla replied:
As far as I understand, auto-focusing is always intrusive from an accessibility point of view. I might be wrong, of course - @diegohaz would know best.

diegohaz replied:
Auto-focus is okay if followed by a user action that makes sense. For example, clicking on a `<label>` should focus on the input; opening a dialog should move focus into it; deleting the currently focused row in a table could move focus somewhere else (like the next or previous rows).

In this component, I would expect that the input would be focused after clicking on the search icon.

tyxla reviewed (commented):
Nice work @saramarcondes.

I'm definitely interested to see a simplification of this component, but it's probably best to be done in another PR. Maybe we could write some tests and then do the refactor.

Yet, I wonder if refactoring it as uncontrolled makes sense - are we going to keep all the functionality that it already offers? 

I've left some thoughts, but this generally looks like pretty solid work!

jsnajdr commented on packages/search/src/search.tsx:
```diff
@@ -0,0 +1,417 @@
+	hideOpenIcon: boolean;
+	initialValue?: string;
+	inputLabel?: string;
+	isOpen: boolean;
```
For uncontrolled components, i.e., these that are not fully controlled by props, but have internal state, there is a convention to use `defaultValue` + `value` or `defaultIsOpen` + `isOpen`. That sets only the initial state on render. Further modifications to this prop don't have any effect.

@diegohaz recently tweeted about this: https://twitter.com/diegohaz/status/1306173502286974976

jsnajdr commented on packages/search/src/search.tsx:
```diff
@@ -0,0 +1,417 @@
+		hasFocus: this.props.autoFocus,
+	};
+
+	UNSAFE_componentWillReceiveProps( nextProps: Props ) {
```
A new component shouldn't use the legacy method. The new `getDerivedStateFromProps` method is preferred. Or even better, do a functional component with hooks.

saramarcondes replied:
This component cannot actually use `getDerivedStateFromProps` because it's being controlled externally and internally. It specifically needs to respond to changes in `isOpen`. If we turn this into an uncontrolled component then we can remove this logic and replace it with a `defaultIsOpen` as you suggested in another comment. Then consumers that need the component to respond to external `isOpen` may add a `key` prop that depends on their `isOpen` value to force the component to re-mount with the new default value.

saramarcondes commented on packages/search/src/search.tsx:
```diff
@@ -0,0 +1,417 @@
+ */
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+import { Spinner, __experimentalInputControl as InputControl } from '@wordpress/components';
```
Originally I used `InputControl` because I didn't notice `TextControl`. However, I'm seeing that `TextControl` does not forward refs, making the logic dependent on the `searchInput` ref not work anymore. I'm testing to see if it's necessary in the first place.

saramarcondes replied:
It turns out the forwarded ref is definitely necessary. I've opened a PR to add that to `TextControl`, though I think I would rather advocate for adopting the experimental `InputControl` which seems poised to eventually replace `TextControl` anyway.

saramarcondes commented on packages/search/src/search.tsx:
```diff
@@ -0,0 +1,417 @@
+/**
+ * Internal variables
+ */
+const SEARCH_DEBOUNCE_MS = 300;
```
This is just used as a default for the `delayTimeout` prop. Did you mean something other than a prop?

tyxla replied:
Disregard, your point makes sense.

saramarcondes commented on packages/search/src/search.tsx:
```diff
@@ -0,0 +1,417 @@
+	isOpen: boolean;
+	maxLength?: number;
+	minLength?: number;
+	onBlur?: ( event: MouseEvent< HTMLInputElement > ) => void;
```
Yes, except it should be `FocusEvent` 🙂

tyxla replied:
Even better! Lol

sarayourfriend committed:
Use TextControl

Also add a noop wrapping form to cause soft-keyboards in iOS to display a "search" button instead of "return": https://stackoverflow.com/a/26287843

diegohaz commented:
Maybe it's out of the scope of this PR since this is just a code extraction. But I wonder why the buttons are `<div>`s and not `<button>`s?

saramarcondes commented:
> But I wonder why the buttons are <div>s and not <button>s?

I think its within scope for this PR. I'd like to extract and make significant improvements wherever possible! I'll update that, thanks for pointing it out. (Plus we can use WordPress `Button`... yet another win 😀)

sarayourfriend committed:
Use actual button instead of div and fix compact icon placement

sarayourfriend committed:
Remove autoFocus from example

sarayourfriend committed:
Fix type errors

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@saramarcondes please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/927651). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

sarayourfriend committed:
Use a regular `input` to allow for stable refs

sarayourfriend committed:
Convert to uncontrolled component

sarayourfriend committed:
Remove unnecessary null checks

sarayourfriend committed:
Clean up and add overlay styling example

a8ci18n commented on packages/search/src/search.tsx:
```diff
@@ -0,0 +1,424 @@
+				tabIndex={ enableOpenIcon ? 0 : undefined }
+				onKeyDown={ enableOpenIcon ? this.openListener : undefined }
+				aria-controls={ 'search-component-' + this.instanceId }
+				aria-label={ __( 'Open Search' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **31** times:
`translate( 'Open Search', { context: 'button label'} )` *ES Score: 14*

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@saramarcondes please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/927727). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

sarayourfriend committed:
Fix rtl value

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@saramarcondes please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/927872). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

sarayourfriend committed:
Remove unnecessary extra element

tyxla commented on packages/search/package.json:
```diff
@@ -0,0 +1,52 @@
+	},
+	"devDependencies": {
+		"@storybook/addon-actions": "^5.3.18",
+		"enzyme": "^3.11.0"
```
Since we don't have any tests yet, it seems we can remove `enzyme` from the list of dependencies, no?

tyxla commented on packages/search/package.json:
```diff
@@ -0,0 +1,52 @@
+{
+	"name": "@automattic/search",
+	"version": "1.0.0-alpha.3",
```
Nit: we haven't released that yet, should it be `1.0.0-alpha` or `beta` or even just `1.0.0`?

tyxla commented on packages/search/src/style.scss:
```diff
@@ -0,0 +1,253 @@
+	}
+}
+
+.search__input-fade .search__text-overlay {
```
I wonder if it makes sense to have all the styles nested in `.search`, to ensure we're not bleeding any styles beyond the component.

Also, while on that topic, I wonder if it makes sense to change the wrapper `.search` to a more specific classname. `.search` seems a bit too common, and IMHO there's a big chance to run into conflicts.

tyxla commented on packages/search/CHANGELOG.md:
```diff
@@ -0,0 +1,3 @@
+## 1.0.0
```
Let's make sure this corresponds to the version in `package.json`.

tyxla commented on packages/search/CHANGELOG.md:
```diff
@@ -0,0 +1,3 @@
+## 1.0.0
+
+- Add Search component
```
Should we mention that this based on the existing code in Calypso?

tyxla commented on packages/search/src/search.tsx:
```diff
@@ -0,0 +1,422 @@
+/**
+ * External dependencies
+ */
+import React, { KeyboardEvent, MouseEvent, FocusEvent, FormEvent, ChangeEvent } from 'react';
```
Opinionated nit: would be really great if we can sort the imports, both when they're multiple imports from the same module and multiple imports on multiple lines. Makes it more readable IMHO.

saramarcondes replied:
Shall we enable an ESLint rule for ordered imports?

tyxla replied:
My personal opinion is YES, but maybe now it's not the time. With the new rule that we recently enabled to make all imports relative to the root of Calypso, any change to existing files will be a mess for contributors.

I say let's reconsider doing at some point in the future.

saramarcondes replied:
/cc @scinos

scinos replied:
Plenty of options if we decide go that route https://www.npmjs.com/search?q=eslint%20sort%20import

tyxla reviewed (approved):
Looking really well and testing great! 

I've left some nits, but nothing blocking IMHO.

I think it's already in a great state and we can address any potential improvement in subsequent PRs. 

Great work @saramarcondes!

sarayourfriend committed:
Remove unnecessary deps and fix version

sarayourfriend committed:
Use more specific classnames

sarayourfriend committed:
Sort imports

sarayourfriend committed:
Nest all styles

sarayourfriend committed:
Update CHANGELOG

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@saramarcondes please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/928620). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

jsnajdr commented on packages/search/package.json:
```diff
@@ -0,0 +1,51 @@
+	"scripts": {
+		"clean": "npx rimraf dist && tsc --build --clean",
+		"prepublish": "yarn run clean",
+		"prepare": "transpile && tsc && copy-assets"
```
Hi @saramarcondes 👋 I noticed that even though the `search` package is written fully in TypeScript, the transpilation for the `dist` directories is done by Babel. The TypeScript compiler is used only to emit the type declaration.

Was this an intentional decision or is it some copy-paste accident? The compilation could be done 100% by the TypeScript compiler, with a pair of `tsconfig.json` and `tsconfig-cjs.json` config files. The `@automattic/data-stores` package does that, for example.

saramarcondes replied:
This is an accident of copy and paste from the way the `components` package is configured. Please feel free to open a PR to fix this if you can, I’m not really familiar with how it would work without Babel...

Side note: the language picker package probably has the same problem.

jsnajdr replied:
Thanks, I'll patch both packages in a PR :+1:
